src/output/plugins/visualization/VisualizationServer.cxx

```diff
@@ -0,0 +1,113 @@
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include "VisualizationServer.hxx"
+
+#include "Log.hxx"
+#include "config/Block.hxx"
+#include "util/Domain.hxx"
+
+using std::make_unique, std::move;
+
+const Domain vis_server_domain("vis_server");
+
+VisualizationServer::VisualizationServer(EventLoop &event_loop,
+                                         const char *bind_to_address,
+                                         uint16_t port,
+                                         size_t max_clients_in,
+                                         const SoundAnalysisParameters &params_in)
+: ServerSocket(event_loop),
+  max_clients(max_clients_in),
+  reaper(event_loop, BIND_THIS_METHOD(ReapClients)),
+  sound_params(params_in)
+{
+	FmtInfo(vis_server_domain,
+            "VisualizationServer::VisualizationServer({}:{}, {} clients maximum)",
+            bind_to_address, port, max_clients);
+
+	ServerSocketAddGeneric(*this, bind_to_address, port);
+}
+
+void VisualizationServer::ReapClients() noexcept
+{
+	FmtNotice(vis_server_domain, "VisualizationServer::ReapClients({}, "
+			  "{} clients)", gettid(), clients.size());
+
+	for (auto p0 = clients.begin(), p1 = clients.end(); p0 != p1; ) {
+		auto p = p0++;
+		if ((*p)->IsClosed()) {
+			LogInfo(vis_server_domain, "Reaping closed client.");
+			clients.erase(p);
+		}
+	}
+
+	if (!clients.empty()) {
+		LogInfo(vis_server_domain, "Scheduling another reaping in 3 seconds.");
+		reaper.Schedule(std::chrono::seconds(3));
+	}
+}
+
+void VisualizationServer::OnPluginOpened(const std::shared_ptr<SoundInfoCache> &pcache)
+{
+    state = HavePcmData{pcache };
+
+	for (auto p0 = clients.begin(), p1 = clients.end(); p0 != p1; ) {
+		auto p = p0++;
+		if (! (*p)->IsClosed()) {
+            (*p)->OnPluginOpened(pcache);
+		}
+	}
+}
+
+void VisualizationServer::OnPluginClosed()
+{
+    state = std::monostate {};
+
+	for (auto p0 = clients.begin(), p1 = clients.end(); p0 != p1; ) {
+		auto p = p0++;
+		if (! (*p)->IsClosed()) {
+            (*p)->OnPluginClosed();
+		}
+	}
+    
+}
+
+void VisualizationServer::OnAccept(UniqueSocketDescriptor fd,
+                                   SocketAddress /*address*/,
+                                   int) noexcept
+{
+	FmtInfo(vis_server_domain, "VisualizationServer::OnAccept({})", gettid());
+
+    // Can we allow an additional client?
+	if (max_clients && clients.size() > max_clients) {
```

Comment:
Off-by-one error? This allows `max_clients+1` clients.

Reply:
Yes-- fixed.